### PR TITLE
Port run_wml_tests to Python and add batch support (part of #4535)

### PR DESCRIPTION
**Pull request by stevecotton**
The verbosity settings are a bit different, but other than that it
should be a drop-in replacement for the old shell script.

Wesnoth's --validcache option isn't used any more, which is part of
the cleanup for #4587 (it didn't speed anything up, because all of
the calculation for the checksum was still done).

@CelticMinstrel please would you test the xcode support? I think
I've implemented it correctly, but don't have a Mac to test on.

### CODE REVIEW

**Comment by stevecotton** [comment]
@AI0867, we both know the next feature that will be added to this, so please would you review this with the assumption that I'll drop the maintenance of it on you? ;-)

**Comment by CelticMinstrel** [comment]
I don't have a Mac to test on either (my Mac is not capable of building Wesnoth 1.13 and greater), so you'll have to ask @hrubymar10 to test it instead.

I can still review the code, though.

**Comment by Pentarctagon** [comment]
Perhaps not as part of this PR, but having one of the macOS jobs run the WML unit tests wouldn't be terrible either, assuming the image travis uses has a recent enough version of python available.

**Comment by CelticMinstrel** [comment]
The code looks fine to me.

**Comment by stevecotton** [comment]
> my Mac is not capable of building Wesnoth 1.13 and greater

That got me thinking about backporting this change. Backporting to 1.12 isn't possible, as it didn't have the WML unit test suite.

Backporting to 1.14 can be easily done:
* delete a few lines in the script's `test_batcher` to yield each test as a separate batch
* handle the extra feature of a793bce969429c4c7993e6a5f96e6f36f06b1489

The extra feature of a793bce969429c4c7993e6a5f96e6f36f06b1489 is to run a test with strict mode disabled, because they pass but log a warning. The change in #4599 is a different method of handling these tests.

File: run_wml_tests [hunk 1]
```diff
@@ -1,391 +1,214 @@
...
+            self.common_args.append("--log-strict=warning")
+        if options.clean:
+            self.common_args.append("--noaddons")
+        if options.additional_arg is not None:
+            self.common_args.extend(options.additional_arg)
+        self.timeout = options.timeout
+        if self.verbose > 1:
+            print("Options that will be used for all Wesnoth instances:", repr(self.common_args))
+
+    def _run_with_rerun_for_sdl_video(args, timeout):
```
Comment (AI0867) [inline run_wml_tests]: This function appears to be unused.

Comment (stevecotton) [reply run_wml_tests]: Thanks for catching this, now fixed (although the function has moved up to file-scope).

**Comment by Pentarctagon** [comment]
It doesn't seem to be a drop-in replacement: https://travis-ci.com/wesnoth/wesnoth/jobs/261064537#L509-L521

**Comment by stevecotton** [comment]
Thanks, the next push will fix the Python 3.5 compatibility.

**Comment by stevecotton** [comment]
Ah, Python 3.5 is the version in Ubuntu 16.04 LTS, but python3 isn't installed in our 16.04 Docker images. @Pentarctagon , would it be okay to add it?

**Comment by Pentarctagon** [comment]
I've pushed out a new image and restarted the failed jobs.  It looks like not all the WML unit tests are being run though?

**Comment by AI0867** [comment]
If a test in a batch fails, does this actually tell us which test it is?

**Comment by stevecotton** [comment]
Thanks for the comments. Fixes in the latest commit:
* list all the tests run, instead of just saying "Running 163 tests"
* show Wesnoth's output when batches of tests fail

```
$ cat wml_test_schedule_with_fails 
# Will be run individually and pass
1 test_return_fail
# Will be run individually and fail (returns PASS when the expectation is FAIL)
1 test_return
# Will be run as a batch of four in this order, and fail on the third one
0 test_return
0 test_return
0 test_return_fail
0 test_return

$ ./run_wml_tests -c '*bunch of path options omitted*' -l wml_test_schedule_with_fails
Running test test_return_fail
Running test test_return
Battle for Wesnoth v1.15.2+dev
Started on Thu Nov 28 21:00:20 2019

Automatically found a possible data directory at BUILD_DIR
Overriding data directory with BUILD_DIR

Data directory:               BUILD_DIR
User configuration directory: /home/USER/.config/wesnoth1.15/
User data directory:          /home/USER/.local/share/wesnoth/1.15
Cache directory:              /home/USER/.cache/wesnoth

Checking lua scripts... ok
PASS TEST: test_return

Failure, Wesnoth returned 0 but we expected 1
Running 4 tests (test_return, test_return, test_return_fail, test_return)
Battle for Wesnoth v1.15.2+dev
Started on Thu Nov 28 21:00:24 2019

Automatically found a possible data directory at BUILD_DIR
Overriding data directory with BUILD_DIR

Data directory:               BUILD_DIR
User configuration directory: /home/USER/.config/wesnoth1.15/
User data directory:          /home/USER/.local/share/wesnoth/1.15
Cache directory:              /home/USER/.cache/wesnoth

Checking lua scripts... ok
PASS TEST: test_return
PASS TEST: test_return
FAIL TEST: test_return_fail

Failure, Wesnoth returned 1 but we expected 0
```

**Pull request merged**